File: backend/providers/sentry.py
"""
Sentry Evidence Provider
Retrieves unhandled exceptions, issue groups, and stack traces from Sentry API.
Provides graceful demo fallback if Sentry credentials are absent or if rate-limited.
"""

import os
import json
import urllib.request
import urllib.error
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from models import (
    EvidenceEvent,
    EvidenceSource,
    ConnectorStatus,
    AcquisitionResult,
    SentryAcquisitionRequest,
)
from .base import EvidenceProvider
from .demo import DemoProvider


class SentryProvider(EvidenceProvider):
    """
    Acquires application runtime error and exception telemetry from Sentry.
    Isolated service with strict timeout handling and fallback guarantee.
    """

    SENTRY_API_BASE = "https://sentry.io/api/0"
    TIMEOUT_SECONDS = 7

    def __init__(self, token: Optional[str] = None):
        self._explicit_token = token

    @property
    def name(self) -> str:
        return "sentry"

    def _get_token(self, request_token: Optional[str] = None) -> Optional[str]:
        return request_token or self._explicit_token or os.getenv("SENTRY_AUTH_TOKEN")

    def _get_org(self, request_org: Optional[str] = None) -> Optional[str]:
        return request_org or os.getenv("SENTRY_ORG")

    def _get_project(self, request_project: Optional[str] = None) -> Optional[str]:
        return request_project or os.getenv("SENTRY_PROJECT")

    async def get_status(
        self,
        token_override: Optional[str] = None,
        org_override: Optional[str] = None,
        project_override: Optional[str] = None,
    ) -> ConnectorStatus:
        token = self._get_token(token_override)
        org = self._get_org(org_override)
        project = self._get_project(project_override)

        if not token:
            return ConnectorStatus(
                connected=False,
                available=True,
                error="SENTRY_AUTH_TOKEN not configured in environment or request",
                fallback_available=True,
                details={"auth_method": "none", "hint": "Configure SENTRY_AUTH_TOKEN or enter token in modal"},
            )

        if not org:
            return ConnectorStatus(
                connected=False,
                available=True,
                error="SENTRY_ORG not configured (organization slug required)",
                fallback_available=True,
            )

        # Probe Sentry project endpoint
        target_project = project or "primary"
        try:
            req = urllib.request.Request(
                f"{self.SENTRY_API_BASE}/projects/{org}/{target_project}/",
                headers={
                    "Authorization": f"Bearer {token}",
                    "User-Agent": "BlackBox-Incident-Investigator/1.0",
                },
            )
            with urllib.request.urlopen(req, timeout=self.TIMEOUT_SECONDS) as resp:
                if resp.status == 200:
                    return ConnectorStatus(
                        connected=True,
                        available=True,
                        error=None,
                        fallback_available=True,
                        details={"organization": org, "project": target_project, "authenticated": True},
                    )
        except urllib.error.HTTPError as e:
            return ConnectorStatus(
                connected=False,
                available=True,
                error=f"Sentry API returned HTTP {e.code}: {e.reason}",
                fallback_available=True,
                details={"http_code": e.code},
            )
        except Exception as e:
            return ConnectorStatus(
                connected=False,
                available=True,
                error=f"Sentry connectivity check failed: {str(e)}",
                fallback_available=True,
            )

        return ConnectorStatus(connected=False, available=True, fallback_available=True)

    async def collect_evidence(
        self,
        organization: Optional[str] = None,
        project: Optional[str] = None,
        incident_start_time: Optional[str] = None,
        incident_end_time: Optional[str] = None,
        query: Optional[str] = None,
        auth_token: Optional[str] = None,
        **kwargs: Any,
    ) -> AcquisitionResult:
        token = self._get_token(auth_token)
        org = self._get_org(organization)
        proj = self._get_project(project) or "payments-service"

        events: List[EvidenceEvent] = []
        warnings: List[str] = []
        used_fallback = False
        error_message: Optional[str] = None

        if not token or not org:
            used_fallback = True
            error_message = "Missing Sentry credentials; loaded authentic demo exception telemetry."
            events = await self._generate_fallback_events(proj)
        else:
            try:
                raw_issues = self._fetch_issues(org, proj, query, token)
                for issue in raw_issues:
                    ev = self._normalize_sentry_issue(issue, org, proj)
                    if ev:
                        events.append(ev)

                if not events:
                    warnings.append(f"No active Sentry issues returned for {org}/{proj}. Loading baseline demo exceptions.")
                    events = await self._generate_fallback_events(proj)
                    used_fallback = True

            except urllib.error.HTTPError as e:
                used_fallback = True
                error_message = f"Sentry API returned HTTP {e.code} ({e.reason}). Activated fallback demo evidence."
                events = await self._generate_fallback_events(proj)
            except Exception as e:
                used_fallback = True
                error_message = f"Sentry connection failed: {str(e)}. Activated fallback demo evidence."
                events = await self._generate_fallback_events(proj)

        status = ConnectorStatus(
            connected=not used_fallback,
            available=True,
            error=error_message,
            fallback_available=True,
            details={"organization": org or "demo-org", "project": proj, "fallback_used": used_fallback},
        )

        return AcquisitionResult(
            source="sentry",
            events_count=len(events),
            sources_count=1,
            summary_message=f"Collected {len(events)} evidence events from 1 source (Sentry).",
            events=events,
            connector_status=status,
            warnings=warnings,
        )

    def _fetch_issues(
        self, org: str, proj: str, query: Optional[str], token: str
    ) -> List[Dict[str, Any]]:
        url = f"{self.SENTRY_API_BASE}/projects/{org}/{proj}/issues/?limit=15"
        if query:
            url += f"&query={urllib.parse.quote(query)}"

        req = urllib.request.Request(
            url,
            headers={
                "Authorization": f"Bearer {token}",
                "User-Agent": "BlackBox-Incident-Investigator/1.0",
                "Accept": "application/json",
            },
        )
        with urllib.request.urlopen(req, timeout=self.TIMEOUT_SECONDS) as resp:
            return json.loads(resp.read().decode("utf-8"))

    def _normalize_sentry_issue(self, raw: Dict[str, Any], org: str, proj: str) -> Optional[EvidenceEvent]:
        issue_id = raw.get("id", "")
        title = raw.get("title", "Unhandled Exception")
        culprit = raw.get("culprit", "")
        count = raw.get("count", "1")
        user_count = raw.get("userCount", 1)
        level = raw.get("level", "error")
        permalink = raw.get("permalink")
        last_seen = raw.get("lastSeen") or datetime.now(timezone.utc).isoformat()
        metadata = raw.get("metadata", {})

        severity = "critical" if level in ["fatal", "critical"] else "error"
        is_candidate = any(
            k in title.lower() for k in ["timeout", "exhaustion", "connection", "memory", "deadlock", "oom", "crash"]
        )

        return EvidenceEvent(
            id=f"evt-sentry-{issue_id}",
            source=EvidenceSource.SENTRY,
            event_type="error",
            timestamp=last_seen,
            title=f"Sentry Issue #{issue_id}: {title}",
            message=culprit or str(metadata.get("value", title)),
            summary=f"Sentry reported {level.upper()} exception ({count} occurrences across {user_count} users): {title}",
            severity=severity,
            metadata={
                "issue_id": issue_id,
                "culprit": culprit,
                "occurrences": count,
                "users_affected": user_count,
                "level": level,
                "organization": org,
                "project": proj,
                "stack_trace": metadata.get("value") or culprit or "Stack trace captured by Sentry agent.",
            },
            related_identifiers=[f"sentry-{issue_id}", culprit, proj],
            raw_reference=permalink,
            is_root_cause_candidate=is_candidate,
        )

    async def _generate_fallback_events(self, proj: str) -> List[EvidenceEvent]:
        demo_provider = DemoProvider()
        res = await demo_provider.collect_evidence(incident_id="inc-db-pool-exhaustion", filter_source="sentry")
        events = res.events
        for e in events:
            e.metadata["is_fallback"] = True
            e.metadata["fallback_project"] = proj
        return events

File: src/services/providers/sentry.ts
import {
  EvidenceEvent,
  ConnectorStatus,
  AcquisitionResult,
  SentryAcquisitionRequest,
} from '../../types/incident';
import { EvidenceProvider } from './base';
import { DemoProvider } from './demo';

export class SentryProvider implements EvidenceProvider {
  name = 'sentry';
  private apiBase = 'https://sentry.io/api/0';
  private timeoutMs = 7000;

  private getToken(requestToken?: string): string | undefined {
    return requestToken || (typeof process !== 'undefined' ? process.env?.SENTRY_AUTH_TOKEN : undefined);
  }

  private getOrg(requestOrg?: string): string | undefined {
    return requestOrg || (typeof process !== 'undefined' ? process.env?.SENTRY_ORG : undefined);
  }

  private getProject(requestProject?: string): string | undefined {
    return requestProject || (typeof process !== 'undefined' ? process.env?.SENTRY_PROJECT : undefined);
  }

  async getStatus(tokenOverride?: string, orgOverride?: string, projOverride?: string): Promise<ConnectorStatus> {
    const token = this.getToken(tokenOverride);
    const org = this.getOrg(orgOverride);
    const project = this.getProject(projOverride) || 'primary';

    if (!token) {
      return {
        connected: false,
        available: true,
        error: 'SENTRY_AUTH_TOKEN not configured in environment or request',
        fallback_available: true,
        details: {
          auth_method: 'none',
          hint: 'Configure SENTRY_AUTH_TOKEN or enter token in connection panel',
        },
      };
    }

    if (!org) {
      return {
        connected: false,
        available: true,
        error: 'SENTRY_ORG not configured (organization slug required)',
        fallback_available: true,
      };
    }

    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

      const resp = await fetch(`${this.apiBase}/projects/${org}/${project}/`, {
        headers: {
          Authorization: `Bearer ${token}`,
          'User-Agent': 'BlackBox-Incident-Investigator/1.0',
        },
        signal: controller.signal,
      });
      clearTimeout(timeout);

      if (resp.ok) {
        return {
          connected: true,
          available: true,
          error: null,
          fallback_available: true,
          details: { organization: org, project, authenticated: true },
        };
      } else {
        return {
          connected: false,
          available: true,
          error: `Sentry API returned HTTP ${resp.status}: ${resp.statusText}`,
          fallback_available: true,
          details: { http_code: resp.status },
        };
      }
    } catch (err: any) {
      return {
        connected: false,
        available: true,
        error: `Sentry probe failed: ${err.message || String(err)}`,
        fallback_available: true,
      };
    }
  }

  async collectEvidence(params: SentryAcquisitionRequest): Promise<AcquisitionResult> {
    const token = this.getToken(params.auth_token);
    const org = this.getOrg(params.organization);
    const project = this.getProject(params.project) || 'payments-service';

    const events: EvidenceEvent[] = [];
    const warnings: string[] = [];
    let usedFallback = false;
    let errorMessage: string | null = null;

    if (!token || !org) {
      usedFallback = true;
      errorMessage = 'Missing Sentry credentials; loaded authentic demo exception telemetry.';
      const fallbackEvents = await this.generateFallbackEvents(project);
      events.push(...fallbackEvents);
    } else {
      try {
        const issues = await this.fetchIssues(org, project, params.query, token);
        for (const issue of issues) {
          const ev = this.normalizeSentryIssue(issue, org, project);
          if (ev) events.push(ev);
        }

        if (events.length === 0) {
          warnings.push(`No active Sentry issues returned for ${org}/${project}. Activated baseline fallback demo exceptions.`);
          usedFallback = true;
          const fallbackEvents = await this.generateFallbackEvents(project);
          events.push(...fallbackEvents);
        }
      } catch (err: any) {
        usedFallback = true;
        errorMessage = `Sentry connection failed (${err.message || String(err)}). Activated fallback demo evidence.`;
        const fallbackEvents = await this.generateFallbackEvents(project);
        events.push(...fallbackEvents);
      }
    }

    const status: ConnectorStatus = {
      connected: !usedFallback,
      available: true,
      error: errorMessage,
      fallback_available: true,
      details: { organization: org || 'demo-org', project, fallback_used: usedFallback },
    };

    return {
      source: 'sentry',
      events_count: events.length,
      sources_count: 1,
      summary_message: `Collected ${events.length} evidence events from 1 source (Sentry).`,
      events,
      connector_status: status,
      warnings,
    };
  }

  private async fetchIssues(org: string, project: string, query: string | undefined, token: string): Promise<any[]> {
    let url = `${this.apiBase}/projects/${org}/${project}/issues/?limit=15`;
    if (query) url += `&query=${encodeURIComponent(query)}`;

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

    const resp = await fetch(url, {
      headers: {
        Authorization: `Bearer ${token}`,
        'User-Agent': 'BlackBox-Incident-Investigator/1.0',
        Accept: 'application/json',
      },
      signal: controller.signal,
    });
    clearTimeout(timeout);

    if (!resp.ok) {
      throw new Error(`HTTP ${resp.status}: ${resp.statusText}`);
    }
    return await resp.json();
  }

  private normalizeSentryIssue(raw: any, org: string, project: string): EvidenceEvent | null {
    const issueId = raw.id || '';
    const title = raw.title || 'Unhandled Exception';
    const culprit = raw.culprit || '';
    const count = raw.count || '1';
    const userCount = raw.userCount || 1;
    const level = (raw.level || 'error').toLowerCase();
    const permalink = raw.permalink;
    const lastSeen = raw.lastSeen || new Date().toISOString();
    const metadata = raw.metadata || {};

    const severity = level === 'fatal' || level === 'critical' ? 'critical' : 'error';
    const isCandidate = /timeout|exhaustion|connection|memory|deadlock|oom|crash/i.test(title);

    return {
      id: `evt-sentry-${issueId}`,
      source: 'sentry',
      event_type: 'error',
      timestamp: lastSeen,
      title: `Sentry Issue #${issueId}: ${title}`,
      message: culprit || metadata.value || title,
      summary: `Sentry reported ${level.toUpperCase()} exception (${count} occurrences across ${userCount} users): ${title}`,
      severity,
      metadata: {
        issue_id: issueId,
        culprit,
        occurrences: count,
        users_affected: userCount,
        level,
        organization: org,
        project,
        stack_trace: metadata.value || culprit || 'Stack trace captured by Sentry agent.',
      },
      related_identifiers: [`sentry-${issueId}`, culprit, project],
      raw_reference: permalink,
      correlated_event_ids: [],
      is_root_cause_candidate: isCandidate,
    };
  }

  private async generateFallbackEvents(project: string): Promise<EvidenceEvent[]> {
    const demo = new DemoProvider();
    const res = await demo.collectEvidence({ filter_source: 'sentry' });
    return res.events.map((e) => ({
      ...e,
      metadata: {
        ...e.metadata,
        is_fallback: true,
        fallback_project: project,
      },
    }));
  }
}
